internal/backoff/backoff.go

```diff
@@ -30,49 +30,156 @@ import (
 
 // Strategy defines the methodology for backing off after a grpc connection
 // failure.
-//
 type Strategy interface {
 	// Backoff returns the amount of time to wait before the next retry given
 	// the number of consecutive failures.
 	Backoff(retries int) time.Duration
+	// MinConnectionTimeout returns the minimum amount of time we are willing to
+	// give a connection to complete.
+	MinConnectionTimeout() time.Duration
 }
 
-const (
-	// baseDelay is the amount of time to wait before retrying after the first
+// StrategyBuilder helps build an implementation of the Strategy interface,
+// with setters provided to set individual parameters of the backoff strategy.
+type StrategyBuilder interface {
```

Comment:
This is not very Go-idiomatic.  Why not export the fields in `Exponential` and set them directly?  This is an internal package, so we can do whatever we want with it w.r.t. changing how the defaults are applied.

Reply:
Yes, if the fields of `Exponential` are exposed, we can directly set them from `WithConnectParams()` where all fields are expected.
But in places where only a subset of the fields are to be modified and the rest should continue to have their default values (especially for other internal packages), this is much easier to use. Also, it makes it easier to allow setting any of the fields to their zero value.